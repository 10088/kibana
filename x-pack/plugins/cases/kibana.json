{
  "configPath": ["xpack", "cases"],
  "id": "cases",
  "kibanaVersion": "kibana",
  "requiredPlugins": ["actions", "securitySolution", "features"],
  "optionalPlugins": [
    "spaces",
    "security"
  ],
  "server": true,
  "ui": false,
  "version": "8.0.0"
}
