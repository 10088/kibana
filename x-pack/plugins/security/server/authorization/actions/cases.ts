/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import { isString } from 'lodash';

export class CasesActions {
  private readonly prefix: string;

  constructor(versionNumber: string) {
    this.prefix = `cases:${versionNumber}:`;
  }

  public get(className: string, operation: string): string {
    if (!operation || !isString(operation)) {
      throw new Error('operation is required and must be a string');
    }

    if (!className || !isString(className)) {
      throw new Error('class is required and must be a string');
    }

    return `${this.prefix}${className}/${operation}`;
  }
}
