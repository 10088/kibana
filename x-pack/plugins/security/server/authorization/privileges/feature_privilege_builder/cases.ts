/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import { uniq } from 'lodash';

import type { FeatureKibanaPrivileges, KibanaFeature } from '../../../../../features/server';
import { BaseFeaturePrivilegeBuilder } from './feature_privilege_builder';

const readOperations: string[] = ['get', 'find'];
const writeOperations: string[] = ['create', 'delete', 'update'];
const allOperations: string[] = [...readOperations, ...writeOperations];

export class FeaturePrivilegeCasesBuilder extends BaseFeaturePrivilegeBuilder {
  public getActions(
    privilegeDefinition: FeatureKibanaPrivileges,
    feature: KibanaFeature
  ): string[] {
    const getCasesPrivilege = (operations: string[], consumer: string) =>
      operations.map((operation) => this.actions.cases.get(consumer, operation));

    // TODO: make sure we don't need to add a cases array or flag? to the FeatureKibanaPrivileges
    // I think we'd only need to do that if we wanted a plugin to be able to get permissions for cases from other plugins?
    // I think we only want the plugin to get access to the cases that are created through itself and not allow it to have
    // access to other plugins

    // It may make sense to add a cases field as a flag so plugins have to opt in to getting access to cases
    return uniq([
      ...getCasesPrivilege(allOperations, feature.id),
      ...getCasesPrivilege(readOperations, feature.id),
    ]);
  }
}
