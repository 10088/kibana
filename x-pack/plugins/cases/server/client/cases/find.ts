/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import Boom from '@hapi/boom';
import { SavedObjectsClientContract, Logger } from 'kibana/server';
import { pipe } from 'fp-ts/lib/pipeable';
import { fold } from 'fp-ts/lib/Either';
import { identity } from 'fp-ts/lib/function';

import {
  CasesFindResponse,
  CasesFindRequest,
  CasesFindRequestRt,
  throwErrors,
  caseStatuses,
  CasesFindResponseRt,
} from '../../../common';
import { CaseServiceSetup } from '../../services';
import { createCaseError } from '../../common/error';
import { constructQueryOptions } from '../../routes/api/cases/helpers';
import { transformCases } from '../../routes/api/utils';

interface FindParams {
  savedObjectsClient: SavedObjectsClientContract;
  caseService: CaseServiceSetup;
  logger: Logger;
  options: CasesFindRequest;
}

/**
 * Retrieves a case and optionally its comments and sub case comments.
 */
export const find = async ({
  savedObjectsClient,
  caseService,
  logger,
  options,
}: FindParams): Promise<CasesFindResponse> => {
  try {
    const queryParams = pipe(
      CasesFindRequestRt.decode(options),
      fold(throwErrors(Boom.badRequest), identity)
    );

    const queryArgs = {
      tags: queryParams.tags,
      reporters: queryParams.reporters,
      sortByField: queryParams.sortField,
      status: queryParams.status,
      caseType: queryParams.type,
    };

    const caseQueries = constructQueryOptions(queryArgs);
    const cases = await caseService.findCasesGroupedByID({
      client: savedObjectsClient,
      caseOptions: { ...queryParams, ...caseQueries.case },
      subCaseOptions: caseQueries.subCase,
    });

    const [openCases, inProgressCases, closedCases] = await Promise.all([
      ...caseStatuses.map((status) => {
        const statusQuery = constructQueryOptions({ ...queryArgs, status });
        return caseService.findCaseStatusStats({
          client: savedObjectsClient,
          caseOptions: statusQuery.case,
          subCaseOptions: statusQuery.subCase,
        });
      }),
    ]);

    return CasesFindResponseRt.encode(
      transformCases({
        ...cases,
        countOpenCases: openCases,
        countInProgressCases: inProgressCases,
        countClosedCases: closedCases,
        total: cases.casesMap.size,
      })
    );
  } catch (error) {
    throw createCaseError({
      message: `Failed to find cases: ${JSON.stringify(options)}: ${error}`,
      error,
      logger,
    });
  }
};
