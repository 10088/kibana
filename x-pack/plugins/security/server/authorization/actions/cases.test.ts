/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import { CasesActions } from './cases';

const version = '1.0.0-zeta1';

describe('#get', () => {
  it.each`
    operation
    ${null}
    ${undefined}
    ${''}
    ${1}
    ${true}
    ${{}}
  `(`operation of ${JSON.stringify('$operation')}`, ({ operation }) => {
    const actions = new CasesActions(version);
    expect(() => actions.get('class', operation)).toThrowErrorMatchingSnapshot();
  });

  it.each`
    className
    ${null}
    ${undefined}
    ${''}
    ${1}
    ${true}
    ${{}}
  `(`class of ${JSON.stringify('$className')}`, ({ className }) => {
    const actions = new CasesActions(version);
    expect(() => actions.get(className, 'operation')).toThrowErrorMatchingSnapshot();
  });

  it('returns `cases:${class}/${operation}`', () => {
    const alertingActions = new CasesActions(version);
    expect(alertingActions.get('security', 'bar-operation')).toBe(
      'cases:1.0.0-zeta1:security/bar-operation'
    );
  });
});
