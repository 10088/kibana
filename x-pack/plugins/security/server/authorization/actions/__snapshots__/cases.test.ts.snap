// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#get class of "" 1`] = `"class is required and must be a string"`;

exports[`#get class of "{}" 1`] = `"class is required and must be a string"`;

exports[`#get class of "1" 1`] = `"class is required and must be a string"`;

exports[`#get class of "null" 1`] = `"class is required and must be a string"`;

exports[`#get class of "true" 1`] = `"class is required and must be a string"`;

exports[`#get class of "undefined" 1`] = `"class is required and must be a string"`;

exports[`#get operation of "" 1`] = `"operation is required and must be a string"`;

exports[`#get operation of "{}" 1`] = `"operation is required and must be a string"`;

exports[`#get operation of "1" 1`] = `"operation is required and must be a string"`;

exports[`#get operation of "null" 1`] = `"operation is required and must be a string"`;

exports[`#get operation of "true" 1`] = `"operation is required and must be a string"`;

exports[`#get operation of "undefined" 1`] = `"operation is required and must be a string"`;
